backend/hidrive/helpers.go

```diff
@@ -0,0 +1,902 @@
+// The io.ReadSeeker should be resettable by seeking to its start.
+// If modTime is not the zero time instant,
+// it will be set as the file's modification time after the operation.
+//
+// This returns fs.ErrorDirNotFound
+// if the parent directory of the file is not found.
+func (f *Fs) overwriteFile(ctx context.Context, path string, content io.ReadSeeker, modTime time.Time) (*api.HiDriveObject, error) {
+	parameters := api.NewQueryParameters()
+	parameters.SetFileInDirectory(path)
+
+	var err error
+	if !modTime.IsZero() {
+		err = parameters.SetTime("mtime", modTime)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	opts := rest.Opts{
+		Method:      "PUT",
+		Path:        "/file",
+		Body:        content,
+		ContentType: "application/octet-stream",
+		Parameters:  parameters.Values,
+	}
+
+	var result api.HiDriveObject
+	var resp *http.Response
+	err = f.pacer.Call(func() (bool, error) {
+		// Reset the reading index (in case this is a retry).
+		if _, err = content.Seek(0, io.SeekStart); err != nil {
+			return false, err
+		}
+		resp, err = f.srv.CallJSON(ctx, &opts, nil, &result)
+		return f.shouldRetry(ctx, resp, err)
+	})
+
+	switch {
+	case err == nil:
+		return &result, nil
+	case isHTTPError(err, 404):
+		return nil, fs.ErrorDirNotFound
+	}
+	return nil, err
+}
+
+// uploadFileChunked updates the content of the existing file at the given path
+// with the content of the io.Reader.
+// Returns the amount of successfully written bytes.
+// If nothing was written this will be 0.
+// Returns the resulting api-object if successful.
+//
+// Replaces the file contents by uploading multiple chunks of the given size in parallel.
+// Therefore this can and be used to upload files of any size efficiently.
+// The number of parallel transfers is limited by transferLimit which should larger than 0.
+// If modTime is not the zero time instant,
+// it will be set as the file's modification time after the operation.
+//
+// NOTE: This method uses updateFileChunked and may create sparse files,
+// if the upload of a chunk fails unexpectedly.
+// See note about sparse files in patchFile.
+// If any of the uploads fail, the process will be aborted and
+// the first error that occurred will be returned.
+// This is not an atomic operation,
+// therefore if the upload fails the file may be partially modified.
+//
+// This returns fs.ErrorObjectNotFound if the object is not found.
+func (f *Fs) uploadFileChunked(ctx context.Context, path string, content io.Reader, modTime time.Time, chunkSize int, transferLimit int64) (amount uint64, info *api.HiDriveObject, err error) {
+	amount, err = f.updateFileChunked(ctx, path, content, 0, chunkSize, transferLimit)
+
+	if err == nil {
+		info, err = f.resizeFile(ctx, path, amount, modTime)
+	}
+	return amount, info, err
+}
+
+// updateFileChunked updates the content of the existing file at the given path
+// starting at the given offset.
+// Returns the amount of successfully written bytes.
```

Comment:
I wasn't quite sure what `amount` is

- Is this just a counter of number of successfully written bytes? If so then using `sync/atomic.AddUint64` will simplify this code a lot.
- If it is supposed to be the last successfully written block then I think it is probably being calculated incorrectly - see below.
- Or is is the end position of the last success write?

I suspect from looking at the code it is supposed to be the last, since its only use seems to be passed into truncate. Maybe it would be best renamed `lastPosition` or something like that if so?

Reply:
It is supposed to be the position of the last successful write, so that all the content before this position is guaranteed to be correct. Using your previous example
> What if block 1 was OK, block 2 failed and block 3 was OK.

`amount` should be the end position of block 1.

Reply:
> amount of successfully written bytes

is a bit misleading in that regard, as one could understand it to include the bytes from block 3. I'll try to clarify it more.

`Sucessful` in this case means that there are no failures up until that position; everything after the first failed write is declared unsuccessful.

Reply:
I'm not sure which of your options fits my description best (it's not 1), but I don't think `lastPosition` makes it clearer. But there may be another more fitting name?